src/python/zquantum/qcbm/target_thermal_states.py

```diff
@@ -0,0 +1,235 @@
+                else:
+                    continue
+        boltzmann_factor = np.exp(energy * beta)
+        partition_function += boltzmann_factor
+
+        binary_bitstring = convert_tuples_to_bitstrings([dec2bin(spin, n_spins)])[-1]
+        distribution[binary_bitstring] = boltzmann_factor
+
+    normalized_distribution = {
+        key: value / partition_function for key, value in distribution.items()
+    }
+
+    return normalized_distribution
+
+
+def get_target_bitstring_distribution(
+    n_spins: int,
+    temperature: float,
+    hamiltonian_parameters: typing.List[np.array],  # TODO: add docstring
+) -> BitstringDistribution:
+    """Generates thermal states target data as BitStringDistribution object
+
+    Args:
+        n_spins: positive number of spins in the Ising system
+        temperature: temperature factor in the boltzman distribution
+
+    Returns:
+       Thermal target Bistring Distribution.
+    """
+    return BitstringDistribution(
+        get_thermal_target_distribution_dict(
+            n_spins, temperature, hamiltonian_parameters
+        )
+    )
+
+
+def cumulate(
+    n_spins: int,
+    temperature: float,
+    hamiltonian_parameters: typing.List[np.array],  # TODO: add docstring
+) -> typing.List[float]:
+    """Generates array of cumulative probabilities from target_distribution.
+    Args:
+        n_spins: positive number of spins in the Ising system
+        temperature: temperature factor in the boltzman distribution
+
+    Returns:
+       List of cumulative probabilities from the target distribution.
+    """
+    probabilities = get_thermal_target_distribution_dict(
+        n_spins, temperature, hamiltonian_parameters
+    )
+    cumulative = []
+    tot = 0.0
+    for i in range(len(probabilities.keys())):
+        tot += list(probabilities.values())[i]
+        cumulative.append(tot)
+    return cumulative
+
+
+# Sample from the inverse cumulative partition function
+def sample(
+    n_samples: int,
+    n_spins: int,
+    temperature: float,
+    hamiltonian_parameters: typing.List[np.array],  # TODO: add docstring
+) -> np.array:
+    """Generates samples from the original target distribution in the form of a list of ising vectors
+    Args:
+        n_samples: the number of samples from the original distribution
+        n_spins: positive number of spins in the Ising system
+        temperature: temperature factor in the boltzman distribution
+    Returns:
+       samples: array of sample n_spin ising vectors
+    """
+    cumulative = cumulate(n_spins, temperature, hamiltonian_parameters)
+    samples = np.zeros((n_samples, n_spins))
+    y = np.zeros(n_samples)
+    for i in range(n_samples):
+        y[i] = np.random.uniform(low=0, high=1)
+    for i in range(len(y)):
+        idx = bisect.bisect_right(cumulative, y[i])
+        ising_bitstring = int2ising(idx, n_spins)
+        samples[i, :] = ising_bitstring
+    return np.asarray(samples)
+
+
+
+def get_thermal_sampled_distribution(
```

Comment:
Same logic as stated with regards to `get_thermal_target_distribution_dict`. I think the code would be cleaner with only one function that returned a `BitstringDistribution` object.

Reply:
Agreed! Done!